File: python/8_StringToInteger.py
def myAtoi(s: str) -> int:
    # find integer from string
    # strip
    ns = s.lstrip()
    # check s[0] && s[1]
    l = len(ns)
    prefix = 1
    if ns[0:1] in ["+", "-"]:
        prefix = -1 if ns[0] == "-" else 1
        ns = ns[1:l]
    i=0
    n = 0
    while (i < len(ns) and ns[i].isnumeric()):
        n = n*10 + int(ns[i])
        i+= 1
    res = prefix * n
    max, min = 2**31-1, -2**31

    if max >= n >= min:
        return res
    else:
        return max if res > max else min

if __name__ == "__main__":
    l = ["    -42", "+-1", "+12", "-01", ".1"]
    for i in l:
        print(myAtoi(i))

File: cpp/8_StringToInteger.cpp
#include <iostream>
#include <string>
#include <math.h>
using namespace std;
int myAtoi(string s) {
    // left trim
    string ns = s.erase(0, s.find_first_not_of(" "));
    int prefix = 1;
    if (ns.length() > 1 && (ns[0] == '+' || ns[0] == '-'))
    {
        prefix = ns[0] == '-' ? -1 : 1;
        ns = ns.substr(1, ns.length());
    }

    long long int n = 0;
    string i;
    long long int max, min, res=0;
    max = pow(2, 31) - 1;
    min = -1 * pow(2, 31);

    for (char& c : ns)
    {
        if (isdigit(c))
        {
            i = c;
            n = n * 10 + stoi(i);
            res = prefix * n;
            // if res is already to overflow stop it immediately
            if (res > max)
                return max;
            else if (res < min)
                return min;
        }
        else
        {
            break;
        }
    }

    return res;
}

int main(){
    return 0;
}
// Use this template always on leetcode to improve runtime.
static const int _ = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
   return 0;
}();

